File: web/src/services/api/influxdbManagement.ts
// InfluxDB 에이전트의 bucket / measurement 관리 API 클라이언트.
//
// 백엔드 REST 계약 (응답은 { data: ... } envelope — apiClient 인터셉터가 벗겨준다):
//   GET    /api/v1/influxdb/{agent_name}/buckets
//     → { buckets: [{ id, name, orgId, retentionSeconds }] }
//   POST   /api/v1/influxdb/{agent_name}/buckets   body { name, retentionSeconds? }
//     → { id, name, orgId, retentionSeconds }
//   DELETE /api/v1/influxdb/{agent_name}/buckets/{bucket}
//     → { deleted: true }
//   POST   /api/v1/influxdb/{agent_name}/buckets/{bucket}/truncate
//     → { truncated: true }
//   GET    /api/v1/influxdb/{agent_name}/measurements?bucket={name}
//     → { measurements: ["m1", "m2"], count }
//   DELETE /api/v1/influxdb/{agent_name}/measurements/{name}?bucket={name}
//     → { deleted: true }
//
// v3 에이전트는 관리 미지원이라 백엔드가 501/400 + 에러 메시지를 반환할 수 있다.
// 이 경우 apiClient 인터셉터가 APIError 로 변환하며, UI 훅/컴포넌트에서 우아하게 처리한다.

import { del, get, post } from './client';

/**
 * InfluxDB 버킷 메타데이터.
 *
 * `retentionSeconds` 는 0 이면 무제한(만료 없음)을 의미한다.
 */
export interface InfluxBucket {
  id: string;
  name: string;
  orgId: string;
  retentionSeconds: number;
}

/** 버킷 생성 요청 본문. `retentionSeconds` 생략 시 백엔드 기본값(무제한)을 사용한다. */
export interface CreateBucketRequest {
  name: string;
  retentionSeconds?: number;
}

/** `GET /buckets` — 버킷 목록을 조회한다. */
export async function fetchInfluxBuckets(
  agentName: string,
): Promise<InfluxBucket[]> {
  // 백엔드는 { buckets: [...] } 객체로 응답한다(envelope 언래핑 후). 이를 배열로 가정하면
  // 소비 측 `.map()` 이 크래시하므로 nested 배열을 추출한다. 구버전/직접 배열 응답과
  // null·빈 객체도 방어적으로 빈 배열로 폴백한다.
  const data = await get<{ buckets?: InfluxBucket[] } | InfluxBucket[] | null>(
    `/influxdb/${encodeURIComponent(agentName)}/buckets`,
  );
  if (Array.isArray(data)) return data;
  return data?.buckets ?? [];
}

/** `POST /buckets` — 버킷을 생성하고 생성된 버킷 메타를 반환한다. */
export async function createInfluxBucket(
  agentName: string,
  req: CreateBucketRequest,
): Promise<InfluxBucket> {
  return post<InfluxBucket>(
    `/influxdb/${encodeURIComponent(agentName)}/buckets`,
    req,
  );
}

/** `DELETE /buckets/{bucket}` — 버킷을 삭제한다. */
export async function deleteInfluxBucket(
  agentName: string,
  bucket: string,
): Promise<void> {
  await del(
    `/influxdb/${encodeURIComponent(agentName)}/buckets/${encodeURIComponent(bucket)}`,
  );
}

/**
 * `POST /buckets/{bucket}/truncate` — 버킷을 초기화한다(데이터 전체 삭제, 버킷은 유지).
 *
 * 파괴적 작업이므로 호출부에서 반드시 강한 확인 절차를 거쳐야 한다.
 */
export async function truncateInfluxBucket(
  agentName: string,
  bucket: string,
): Promise<void> {
  await post<{ truncated: boolean }>(
    `/influxdb/${encodeURIComponent(agentName)}/buckets/${encodeURIComponent(bucket)}/truncate`,
  );
}

/** `GET /measurements?bucket={name}` — 지정 버킷의 measurement 이름 목록을 조회한다. */
export async function fetchInfluxMeasurements(
  agentName: string,
  bucket: string,
): Promise<string[]> {
  const params = new URLSearchParams({ bucket });
  // 백엔드는 { measurements: [...], count } 객체로 응답한다(envelope 언래핑 후).
  // buckets 와 동일하게 nested 배열을 추출하고, 직접 배열/null·빈 객체는 폴백한다.
  const data = await get<{ measurements?: string[] } | string[] | null>(
    `/influxdb/${encodeURIComponent(agentName)}/measurements?${params}`,
  );
  if (Array.isArray(data)) return data;
  return data?.measurements ?? [];
}

/** `DELETE /measurements/{name}?bucket={name}` — 지정 버킷의 measurement 를 삭제한다. */
export async function deleteInfluxMeasurement(
  agentName: string,
  bucket: string,
  measurement: string,
): Promise<void> {
  const params = new URLSearchParams({ bucket });
  await del(
    `/influxdb/${encodeURIComponent(agentName)}/measurements/${encodeURIComponent(measurement)}?${params}`,
  );
}

// ---- 스키마 디스커버리 (SPEC-TSDB-002 §2.10 D2/D3/D4) ----
//
// 백엔드 REST 계약 (M5 에서 신설, `internal/api/handler/influxdb_management.go`):
//   GET /api/v1/influxdb/{agent_name}/tag-keys?measurement=&bucket=
//     → { tag_keys: ["host", "region"], count }
//   GET /api/v1/influxdb/{agent_name}/tag-values?measurement=&tag_key=&bucket=
//     → { tag_values: ["a", "b"], count }
//   GET /api/v1/influxdb/{agent_name}/field-keys?measurement=&bucket=
//     → { field_keys: ["usage", "idle"], count }
//
// 세 라우트는 **관리 조작이 아니라 디스커버리**이므로 v3 에이전트에서도 활성이다
// (§2.10 · §2.13). 관리 조작(생성 · 삭제 · truncate)만 v3 에서 501 이다.
//
// 디스커버리 응답은 캐시하지 않는다(UB1-12) — 스키마는 쓰기와 함께 계속 변하므로
// 캐시된 목록은 방금 생성된 measurement/field 를 감춘다. 캐시 정책은 호출부 소관이며
// 이 모듈은 매 호출을 그대로 네트워크로 보낸다.

/**
 * 디스커버리 응답의 배열 필드를 방어적으로 추출한다.
 *
 * `fetchInfluxBuckets` · `fetchInfluxMeasurements` 와 같은 규칙이다 — 서버가 객체
 * envelope 를 주지만, 구버전/직접 배열 응답과 null 도 빈 배열로 접는다. 소비 측
 * `.map()` 이 크래시하지 않는 것이 이 폴백의 유일한 목적이다.
 */
function pickStringList(
  data: Record<string, unknown> | string[] | null,
  field: string,
): string[] {
  if (Array.isArray(data)) return data;
  const list = data?.[field];
  return Array.isArray(list) ? (list as string[]) : [];
}

/** `GET /tag-keys` — measurement 한정 태그 키 목록(D2). */
export async function fetchInfluxTagKeys(
  agentName: string,
  measurement: string,
  bucket?: string,
): Promise<string[]> {
  const params = new URLSearchParams({ measurement });
  if (bucket) params.set('bucket', bucket);
  const data = await get<Record<string, unknown> | string[] | null>(
    `/influxdb/${encodeURIComponent(agentName)}/tag-keys?${params}`,
  );
  return pickStringList(data, 'tag_keys');
}

/**
 * `GET /tag-values` — 지정 태그 키의 값 목록(D3).
 *
 * `tagKey` 는 필수다. 백엔드도 빈 값을 400 으로 거부한다 — 어느 키의 값인지 모르는
 * 목록은 쓸모가 없기 때문이다.
 */
export async function fetchInfluxTagValues(
  agentName: string,
  measurement: string,
  tagKey: string,
  bucket?: string,
  /**
   * 사전 필터(`k=v`). 값 목록을 그 조건 아래로 좁힌다.
   *
   * 그룹 미리보기가 이 인자를 쓴다 — 시리즈 열거는 접기 전 원시 행 상한에 걸려
   * 고빈도 measurement 에서 값 일부만 주지만, 태그 값 조회는 메타데이터 질의라
   * 그 상한과 무관하다. @spec SPEC-TSDB-004
   */
  filters?: Record<string, string>,
  /**
   * 조회 시간창. **비우면 백엔드의 암묵 기본값이 적용된다** — v3 의
   * SHOW TAG VALUES 는 최근 창만 훑고 Flux 는 -30d 다. 그 창 밖에서만 보고한
   * 장비가 목록에서 조용히 빠지므로, 패널이 그리는 창을 넘겨야 한다.
   * @spec SPEC-TSDB-004 UB1-19
   */
  window?: { startMs: number; endMs: number },
): Promise<string[]> {
  const params = new URLSearchParams({ measurement, tag_key: tagKey });
  if (bucket) params.set('bucket', bucket);
  if (filters && Object.keys(filters).length > 0) {
    params.set(
      'tags',
      Object.keys(filters)
        .sort()
        .map((k) => `${k}=${filters[k]}`)
        .join(','),
    );
  }
  if (window) {
    params.set('start_ms', String(window.startMs));
    params.set('end_ms', String(window.endMs));
  }
  const data = await get<Record<string, unknown> | string[] | null>(
    `/influxdb/${encodeURIComponent(agentName)}/tag-values?${params}`,
  );
  return pickStringList(data, 'tag_values');
}

/** `GET /field-keys` — measurement 한정 필드 키 목록(D4). */
export async function fetchInfluxFieldKeys(
  agentName: string,
  measurement: string,
  bucket?: string,
): Promise<string[]> {
  const params = new URLSearchParams({ measurement });
  if (bucket) params.set('bucket', bucket);
  const data = await get<Record<string, unknown> | string[] | null>(
    `/influxdb/${encodeURIComponent(agentName)}/field-keys?${params}`,
  );
  return pickStringList(data, 'field_keys');
}

File: internal/api/handler/influxdb_management.go
package handler

import (
	"context"
	"errors"
	"log/slog"
	"net/http"
	"time"

	"github.com/xtra/xflow/internal/agent/system"
	"github.com/xtra/xflow/internal/api"
	"github.com/xtra/xflow/internal/api/dto"
)

// influxManager 는 InfluxDB 에이전트가 제공해야 하는 bucket/measurement 관리 계약이다.
// *system.InfluxDBAgent 가 이 인터페이스를 만족한다. 잘못된 타입의 에이전트가
// 관리 엔드포인트에 지정되면 400 으로 거부한다.
type influxManager interface {
	ListBuckets(ctx context.Context) ([]system.BucketInfo, error)
	CreateBucket(ctx context.Context, name string, retentionSeconds int64) (system.BucketInfo, error)
	DeleteBucket(ctx context.Context, bucketRef string) error
	TruncateBucket(ctx context.Context, bucketRef string) error
	ListMeasurements(ctx context.Context, bucket string) ([]string, error)
	DeleteMeasurement(ctx context.Context, bucket, measurement string) error
}

// influxSchemaDiscoverer 는 InfluxDB 에이전트가 제공해야 하는 스키마 디스커버리
// 계약이다. *system.InfluxDBAgent 가 이 인터페이스를 만족한다.
//
// @spec SPEC-TSDB-002 §2.10 (U10)
//
// influxManager 와 합치지 않는 이유는 권한 축이 다르기 때문이다 — 디스커버리는
// 읽기 전용(store.read)이고 관리 조작은 쓰기(store.update)를 포함한다. 하나로
// 합치면 읽기 전용 라우트가 쓰기 능력을 갖춘 에이전트만 받게 된다.
type influxSchemaDiscoverer interface {
	ListTagKeys(ctx context.Context, bucket, measurement string) ([]string, error)
	ListTagValues(ctx context.Context, bucket, measurement, tagKey string) ([]string, error)
	ListFieldKeys(ctx context.Context, bucket, measurement string) ([]string, error)
}

// defaultInfluxManagementTimeout 은 관리 조작 HTTP 요청의 기본 타임아웃이다.
// 대용량 truncate/delete 는 서버 측 삭제 작업이 길어질 수 있어 쿼리보다 넉넉하게 잡는다.
var defaultInfluxManagementTimeout = 60 * time.Second

// InfluxDBManagementHandler 는 InfluxDB bucket/measurement 관리 라우트를 제공한다.
//
// 라우트:
//
//	GET    /influxdb/{agent_name}/buckets
//	POST   /influxdb/{agent_name}/buckets
//	DELETE /influxdb/{agent_name}/buckets/{bucket}
//	POST   /influxdb/{agent_name}/buckets/{bucket}/truncate
//	GET    /influxdb/{agent_name}/measurements?bucket=
//	DELETE /influxdb/{agent_name}/measurements/{name}?bucket=
//
// 스키마 디스커버리 라우트(§2.10 D2~D4, 전부 읽기 전용):
//
//	GET    /influxdb/{agent_name}/tag-keys?measurement=&bucket=
//	GET    /influxdb/{agent_name}/tag-values?measurement=&tag_key=&bucket=
//	GET    /influxdb/{agent_name}/field-keys?measurement=&bucket=
//
// D1(measurement 목록)은 위 /measurements 라우트가 겸한다 — 신규 라우트를 만들지
// 않는다. 같은 목록을 두 경로가 돌려주면 어느 쪽이 정본인지 알 수 없게 된다.
type InfluxDBManagementHandler struct {
	agents AgentLookup
	logger *slog.Logger
}

// NewInfluxDBManagementHandler 는 새 InfluxDBManagementHandler 를 생성한다.
func NewInfluxDBManagementHandler(agents AgentLookup, logger *slog.Logger) *InfluxDBManagementHandler {
	if logger == nil {
		logger = slog.Default()
	}
	return &InfluxDBManagementHandler{agents: agents, logger: logger}
}

// RegisterRoutes 는 InfluxDB 관리 라우트를 등록한다.
func (h *InfluxDBManagementHandler) RegisterRoutes(g *api.RouteGroup) {
	// @SPEC:SPEC-AUTH-005 (M5) — 카탈로그에 influxdb 리소스가 없어 store.* 로 매핑한다.
	// store 는 read/update 2종이므로 bucket/measurement 삭제도 update 이다.
	g.GETPerm("/influxdb/{agent_name}/buckets", "store.read", h.ListBuckets)
	g.POSTPerm("/influxdb/{agent_name}/buckets", "store.update", h.CreateBucket)
	g.DELETEPerm("/influxdb/{agent_name}/buckets/{bucket}", "store.update", h.DeleteBucket)
	g.POSTPerm("/influxdb/{agent_name}/buckets/{bucket}/truncate", "store.update", h.TruncateBucket)
	g.GETPerm("/influxdb/{agent_name}/measurements", "store.read", h.ListMeasurements)
	g.DELETEPerm("/influxdb/{agent_name}/measurements/{name}", "store.update", h.DeleteMeasurement)

	// @spec SPEC-TSDB-002 §2.10 (U10) — 스키마 디스커버리 D2~D4.
	// 전부 조회이므로 store.read 이다(D1 /measurements 와 같은 권한).
	g.GETPerm("/influxdb/{agent_name}/tag-keys", "store.read", h.ListTagKeys)
	g.GETPerm("/influxdb/{agent_name}/tag-values", "store.read", h.ListTagValues)
	g.GETPerm("/influxdb/{agent_name}/field-keys", "store.read", h.ListFieldKeys)
}

// resolveDiscoverer 는 agent_name 으로 에이전트를 조회하고 influxSchemaDiscoverer
// 로 검증한다. 판정 실패의 처분은 resolveManager 와 같다(404 / 400).
func (h *InfluxDBManagementHandler) resolveDiscoverer(ctx api.Context) (influxSchemaDiscoverer, *api.APIError) {
	agentName := ctx.Param("agent_name")
	if agentName == "" {
		return nil, api.ErrBadRequest.WithMessage("agent_name is required")
	}

	ag := findAgentByName(h.agents, agentName)
	if ag == nil {
		return nil, api.ErrNotFound.
			WithMessage("agent_not_found: " + agentName).
			WithDetails(map[string]string{"error": "agent_not_found"})
	}

	disc, ok := ag.(influxSchemaDiscoverer)
	if !ok {
		return nil, api.ErrBadRequest.
			WithMessage("not_an_influxdb_agent: " + agentName).
			WithDetails(map[string]string{"error": "not_an_influxdb_agent"})
	}
	return disc, nil
}

// resolveManager 는 agent_name 으로 에이전트를 조회하고 influxManager 로 검증한다.
//   - 에이전트 없음: 404
//   - influxManager 미구현(잘못된 타입): 400
func (h *InfluxDBManagementHandler) resolveManager(ctx api.Context) (influxManager, *api.APIError) {
	agentName := ctx.Param("agent_name")
	if agentName == "" {
		return nil, api.ErrBadRequest.WithMessage("agent_name is required")
	}

	ag := findAgentByName(h.agents, agentName)
	if ag == nil {
		return nil, api.ErrNotFound.
			WithMessage("agent_not_found: " + agentName).
			WithDetails(map[string]string{"error": "agent_not_found"})
	}

	mgr, ok := ag.(influxManager)
	if !ok {
		return nil, api.ErrBadRequest.
			WithMessage("not_an_influxdb_agent: " + agentName).
			WithDetails(map[string]string{"error": "not_an_influxdb_agent"})
	}
	return mgr, nil
}

// mapInfluxManagementError 는 관리 조작 에러를 적절한 HTTP 에러로 매핑한다.
//   - ErrManagementNotSupported → 501 Not Implemented
//   - context 타임아웃/취소 → 408 Request Timeout
//   - 그 외 → 502 Bad Gateway (업스트림 InfluxDB 조작 실패)
func mapInfluxManagementError(err error) *api.APIError {
	switch {
	case errors.Is(err, system.ErrManagementNotSupported):
		return api.ErrNotImplemented.
			WithMessage(err.Error()).
			WithDetails(map[string]string{"error": "management_not_supported"})
	case errors.Is(err, context.DeadlineExceeded), errors.Is(err, context.Canceled):
		return api.ErrRequestTimeout.WithMessage(err.Error())
	default:
		return api.ErrBadGateway.WithMessage(err.Error())
	}
}

// bucketInfoDTO 는 BucketInfo 의 HTTP 응답 표현이다.
type bucketInfoDTO struct {
	ID               string `json:"id"`
	Name             string `json:"name"`
	OrgID            string `json:"org_id"`
	RetentionSeconds int64  `json:"retention_seconds"`
}

func toBucketInfoDTO(b system.BucketInfo) bucketInfoDTO {
	return bucketInfoDTO{
		ID:               b.ID,
		Name:             b.Name,
		OrgID:            b.OrgID,
		RetentionSeconds: b.RetentionSeconds,
	}
}

// createBucketRequest 는 버킷 생성 요청 바디이다.
type createBucketRequest struct {
	Name             string `json:"name"`
	RetentionSeconds int64  `json:"retentionSeconds"`
}

// ListBuckets 는 GET /influxdb/{agent_name}/buckets 를 처리한다.
func (h *InfluxDBManagementHandler) ListBuckets(ctx api.Context) error {
	mgr, apiErr := h.resolveManager(ctx)
	if apiErr != nil {
		return apiErr
	}

	mctx, cancel := context.WithTimeout(ctx.Context(), defaultInfluxManagementTimeout)
	defer cancel()

	buckets, err := mgr.ListBuckets(mctx)
	if err != nil {
		return mapInfluxManagementError(err)
	}

	items := make([]bucketInfoDTO, 0, len(buckets))
	for _, b := range buckets {
		items = append(items, toBucketInfoDTO(b))
	}
	return ctx.JSON(http.StatusOK, dto.NewSuccessResponse(map[string]any{
		"buckets": items,
		"count":   len(items),
	}))
}

// CreateBucket 은 POST /influxdb/{agent_name}/buckets 를 처리한다.
func (h *InfluxDBManagementHandler) CreateBucket(ctx api.Context) error {
	mgr, apiErr := h.resolveManager(ctx)
	if apiErr != nil {
		return apiErr
	}

	var req createBucketRequest
	if err := ctx.Bind(&req); err != nil {
		return err
	}
	if req.Name == "" {
		return api.ErrBadRequest.WithMessage("name is required")
	}
	if req.RetentionSeconds < 0 {
		return api.ErrBadRequest.WithMessage("retentionSeconds must be >= 0")
	}

	mctx, cancel := context.WithTimeout(ctx.Context(), defaultInfluxManagementTimeout)
	defer cancel()

	info, err := mgr.CreateBucket(mctx, req.Name, req.RetentionSeconds)
	if err != nil {
		return mapInfluxManagementError(err)
	}
	return ctx.JSON(http.StatusOK, dto.NewSuccessResponse(toBucketInfoDTO(info)))
}

// DeleteBucket 은 DELETE /influxdb/{agent_name}/buckets/{bucket} 를 처리한다.
func (h *InfluxDBManagementHandler) DeleteBucket(ctx api.Context) error {
	mgr, apiErr := h.resolveManager(ctx)
	if apiErr != nil {
		return apiErr
	}

	bucket := ctx.Param("bucket")
	if bucket == "" {
		return api.ErrBadRequest.WithMessage("bucket is required")
	}

	mctx, cancel := context.WithTimeout(ctx.Context(), defaultInfluxManagementTimeout)
	defer cancel()

	if err := mgr.DeleteBucket(mctx, bucket); err != nil {
		return mapInfluxManagementError(err)
	}
	return ctx.JSON(http.StatusOK, dto.NewSuccessResponse(map[string]any{
		"bucket":  bucket,
		"deleted": true,
	}))
}

// TruncateBucket 은 POST /influxdb/{agent_name}/buckets/{bucket}/truncate 를 처리한다.
func (h *InfluxDBManagementHandler) TruncateBucket(ctx api.Context) error {
	mgr, apiErr := h.resolveManager(ctx)
	if apiErr != nil {
		return apiErr
	}

	bucket := ctx.Param("bucket")
	if bucket == "" {
		return api.ErrBadRequest.WithMessage("bucket is required")
	}

	mctx, cancel := context.WithTimeout(ctx.Context(), defaultInfluxManagementTimeout)
	defer cancel()

	if err := mgr.TruncateBucket(mctx, bucket); err != nil {
		return mapInfluxManagementError(err)
	}
	return ctx.JSON(http.StatusOK, dto.NewSuccessResponse(map[string]any{
		"bucket":    bucket,
		"truncated": true,
	}))
}

// ListMeasurements 는 GET /influxdb/{agent_name}/measurements?bucket= 를 처리한다.
// bucket 쿼리 파라미터를 생략하면 에이전트 설정의 기본 버킷을 사용한다.
func (h *InfluxDBManagementHandler) ListMeasurements(ctx api.Context) error {
	mgr, apiErr := h.resolveManager(ctx)
	if apiErr != nil {
		return apiErr
	}

	bucket := ctx.Query("bucket")

	mctx, cancel := context.WithTimeout(ctx.Context(), defaultInfluxManagementTimeout)
	defer cancel()

	measurements, err := mgr.ListMeasurements(mctx, bucket)
	if err != nil {
		return mapInfluxManagementError(err)
	}
	if measurements == nil {
		measurements = []string{}
	}
	return ctx.JSON(http.StatusOK, dto.NewSuccessResponse(map[string]any{
		"measurements": measurements,
		"count":        len(measurements),
	}))
}

// DeleteMeasurement 는 DELETE /influxdb/{agent_name}/measurements/{name}?bucket= 를 처리한다.
// bucket 쿼리 파라미터를 생략하면 에이전트 설정의 기본 버킷을 사용한다.
func (h *InfluxDBManagementHandler) DeleteMeasurement(ctx api.Context) error {
	mgr, apiErr := h.resolveManager(ctx)
	if apiErr != nil {
		return apiErr
	}

	measurement := ctx.Param("name")
	if measurement == "" {
		return api.ErrBadRequest.WithMessage("measurement name is required")
	}
	bucket := ctx.Query("bucket")

	mctx, cancel := context.WithTimeout(ctx.Context(), defaultInfluxManagementTimeout)
	defer cancel()

	if err := mgr.DeleteMeasurement(mctx, bucket, measurement); err != nil {
		return mapInfluxManagementError(err)
	}
	return ctx.JSON(http.StatusOK, dto.NewSuccessResponse(map[string]any{
		"measurement": measurement,
		"bucket":      bucket,
		"deleted":     true,
	}))
}

// --- 스키마 디스커버리 (@spec SPEC-TSDB-002 §2.10 D2~D4) ---
//
// 세 핸들러 모두 읽기 전용이며 응답을 캐시하지 않는다(UB1-12). 스키마는 쓰기에
// 따라 변하므로, 오래된 목록에서 고른 시리즈는 조회 시 빈 결과가 된다.

// mapInfluxDiscoveryError 는 디스커버리 오류를 HTTP 오류로 매핑한다.
//
// 이스케이프 불가 식별자는 400 이다 — 사용자가 고칠 수 있는 입력 오류를 502 로
// 보내면 클라이언트가 업스트림 장애로 오독한다. 나머지는 관리 조작과 같은
// 매핑(501 / 408 / 502)을 쓴다.
func mapInfluxDiscoveryError(err error) *api.APIError {
	if errors.Is(err, system.ErrUnescapableIdentifier) {
		return api.ErrBadRequest.WithMessage(err.Error())
	}
	return mapInfluxManagementError(err)
}

// ListTagKeys 는 GET /influxdb/{agent_name}/tag-keys?measurement=&bucket= 를 처리한다(D2).
func (h *InfluxDBManagementHandler) ListTagKeys(ctx api.Context) error {
	disc, apiErr := h.resolveDiscoverer(ctx)
	if apiErr != nil {
		return apiErr
	}

	measurement := ctx.Query("measurement")
	if measurement == "" {
		return api.ErrBadRequest.WithMessage("measurement is required")
	}

	mctx, cancel := context.WithTimeout(ctx.Context(), defaultInfluxManagementTimeout)
	defer cancel()

	keys, err := disc.ListTagKeys(mctx, ctx.Query("bucket"), measurement)
	if err != nil {
		return mapInfluxDiscoveryError(err)
	}
	if keys == nil {
		keys = []string{}
	}
	return ctx.JSON(http.StatusOK, dto.NewSuccessResponse(map[string]any{
		"tag_keys": keys,
		"count":    len(keys),
	}))
}

// ListTagValues 는 GET /influxdb/{agent_name}/tag-values?measurement=&tag_key=&bucket=
// 를 처리한다(D3).
func (h *InfluxDBManagementHandler) ListTagValues(ctx api.Context) error {
	disc, apiErr := h.resolveDiscoverer(ctx)
	if apiErr != nil {
		return apiErr
	}

	measurement := ctx.Query("measurement")
	if measurement == "" {
		return api.ErrBadRequest.WithMessage("measurement is required")
	}
	tagKey := ctx.Query("tag_key")
	if tagKey == "" {
		// 태그 키 없이 값 목록을 돌려주면 어느 키의 값인지 알 수 없다. 폴백을
		// 두지 않고 거부한다.
		return api.ErrBadRequest.WithMessage("tag_key is required")
	}

	mctx, cancel := context.WithTimeout(ctx.Context(), defaultInfluxManagementTimeout)
	defer cancel()

	values, err := disc.ListTagValues(mctx, ctx.Query("bucket"), measurement, tagKey)
	if err != nil {
		return mapInfluxDiscoveryError(err)
	}
	if values == nil {
		values = []string{}
	}
	return ctx.JSON(http.StatusOK, dto.NewSuccessResponse(map[string]any{
		"tag_values": values,
		"count":      len(values),
	}))
}

// ListFieldKeys 는 GET /influxdb/{agent_name}/field-keys?measurement=&bucket= 를 처리한다(D4).
func (h *InfluxDBManagementHandler) ListFieldKeys(ctx api.Context) error {
	disc, apiErr := h.resolveDiscoverer(ctx)
	if apiErr != nil {
		return apiErr
	}

	measurement := ctx.Query("measurement")
	if measurement == "" {
		return api.ErrBadRequest.WithMessage("measurement is required")
	}

	mctx, cancel := context.WithTimeout(ctx.Context(), defaultInfluxManagementTimeout)
	defer cancel()

	fields, err := disc.ListFieldKeys(mctx, ctx.Query("bucket"), measurement)
	if err != nil {
		return mapInfluxDiscoveryError(err)
	}
	if fields == nil {
		fields = []string{}
	}
	return ctx.JSON(http.StatusOK, dto.NewSuccessResponse(map[string]any{
		"field_keys": fields,
		"count":      len(fields),
	}))
}
